hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/TypeEquivalance.java

```diff
@@ -41,6 +52,12 @@ public boolean isComparable(Object o) {
 
 	@Override
 	public int compare(Object x, Object y) {
-		return type.compare(x, y);
+		Class<?> typeClass = type.getReturnedClass();
+		if (typeClass.isInstance(x) && typeClass.isInstance(y)) {
+			return type.compare(x, y);
+		}
+		else {
+			return 0;
```

Comment:
Should we throw an exception instead of returning zero?  I noticed that java.util.Comparator.int compare(T o1, T o2) can "throw an ClassCastException if the arguments' types prevent them from being compared by this comparator."